test/stripe/stripe_client_test.rb

```diff
@@ -82,8 +100,64 @@ class StripeClientTest < Test::Unit::TestCase
         assert_equal 1, StripeClient.maybe_gc_connection_managers
 
         # And as an additional check, the connection manager of the current
-        # thread context should have been set to `nil` as it was GCed.
-        assert_nil StripeClient.current_thread_context.default_connection_manager
+        # thread context should have been removed as it was GCed.
+        assert_equal({}, StripeClient.current_thread_context.default_connection_managers)
+      end
+
+      should "only garbage collect when all connection managers for a thread are expired" do
```

Comment:
@brandur-stripe added this spec to capture the behavior of waiting for all connection managers to be expired before GCing a `thread_context`.

Side note: I got bit pretty hard when writing this spec because I created the second `StripeClient` with the config override `{read_timeout: 100}` without realizing that some configuration options will call `StripeClient.clear_all_connection_managers`. I wonder if that's something worth pointing out in the README since it could adversely affect those trying to get bleeding performance?

Reply:
Thanks! And very nice — the connect manager management logic looks great now.

> Side note: I got bit pretty hard when writing this spec because I created the second `StripeClient` with the config override `{read_timeout: 100}` without realizing that some configuration options will call `StripeClient.clear_all_connection_managers`. I wonder if that's something worth pointing out in the README since it could adversely affect those trying to get bleeding performance?

Oof :/ Actually, the fact that this happens now should be considered a bug. The reason that setting those configurations clears config is that we previously only had a single configuration object, and when people set a new global configuration like `Stripe.api_base = `, it was appropriate to clear out any existing clients. Now that we have specific client configurations, the preferable way of doing that instead would be just to create a new `StripeClient` with a different configuration. I suppose we could also support changing configuration per-client, but in that case we should only clear out connect managers relevant to that specific config object.

I'll bring in your PR first and take a stab at fixing this separately.